### DISCUSSION TITLE
V10 Changes Discussion

### DISCUSSION THREAD

**Discussion by marcus-j-davies** [zwave-js/node-red-contrib-zwave-js#259] [Ideas]
@crxporter I'm dragging you down with me 😄 
but anyone is welcome

**Initial Goals**
 - Transition to typescript
    - Better type affinity in config values (and throughout)
    - Less error prone
    - Easier for contributors 
    
 - Migrate to config nodes 
    - Less memory Usage
    - Unlimited Networks
    - Cleaner flows
    
**Assumed Barriers**
  - Controller nodes will break
  - Device nodes will break
   - Documentation/Example updates
    
 This is no walk in the park, and will take a considerable amount of time.

**Comment by crxporter** [comment]
...watch this space...

Like always, I bet you've got good ideas and will do it the right way. 
Hit me up when you're trying to make sure it's not too complicated for the user.

**Comment by fabnavigator** [comment]
I would like to be able to access the cached node status information from within a function node. I want to be able to call a function, pass it some parameters, and have it return with a cached value (i.e., the state of a switch node). Hopefully this makes some sense. As far as I know, to do this today, I would have to send a message to the controller node and wait for a response.

**Comment by marcus-j-davies** [comment]
Hi @fabnavigator,

This is fully supported already - and returns the cached value.

The function is **getValue** - and returns the value that is currently known (it does not poll the device)

The function that does poll the device is **pollValue** - this one you do have to wait for a response but for **getValue** it's instant. 

I'm on a mobile currently, but check the wiki on GitHub for **getValue**.


Link: https://github.com/zwave-js/node-red-contrib-zwave-js/wiki/Value-API#getvalue

**Reply by fabnavigator** [reply]
Hi @marcus-j-davies,

Hmm... I've been creating a payload with the ValueAPI mode and the getValue method, and passing the payload to the ZWave controller. Then I wait for a message from the controller with my value. If there is a JS function that I can call from with a function node, I don't know about that. No rush with this. I know it's late there.

**Reply by crxporter** [reply]
That's not really how nodered works though... running some api from a function node is like a last resort if there's no nodered contrib that can give you that data. 

The nodered way is to use some piec of the zwave contrib to grab the zwave value and pass it to the next node. 

That being said, I've made for myself a cache of sorts in global context when I need some zwave status information often. I just run like `global.get("zwaveDoorState")` for example. I'm using the getDB as Marcus described to populate the globals on startup. 

Not trying to say you're wrong, just trying to add my method to the discussion and wondering a bit about why you'd need such "raw" access to the zwave api from a function node...

**Reply by fabnavigator** [reply]
@crxporter, You use the "global.get" method from a function node. I'm just saying it would be nice to have a zwave specific method (i.e., "zwave.get") that could be called from a function node. I am doing the same thing you are with context to get access to zwave status information from within a function node. I have a multiple door sensor situation that requires node status information.

Thinking this though, I'm probably just looking for someone who knows how to do it, to write an external function that I can use to get the cached node status information.

**Reply by crxporter** [reply]
It seems like piece of zwave js you're looking for is what Marcus suggested you to use, the `getValue` method. You'd supply a node number and value id to the device node then it'll output the latest cached value (it can poll, too, but that wouldn't work with battery devices). 

Zwave js documentation of this method is [here](https://zwave-js.github.io/node-zwave-js/#/api/node?id=zwavenode-methods). 

It would be fairly simple to make a subflow that would help with these calls. If you want some help building this, feel free to start a new discussion and flag me. I should have a good amount of time this week to help.

At this point I don't think anything new is needed in the node but I reserve the right to change my mind after we dig into it more.

**Reply by marcus-j-davies** [reply]
So @fabnavigator, @crxporter 

Something of high interest could come from this input.

The `flow`, `node `& `global `context are not restricted to the flows, they can be set within a custom-built node
(i.e. node-red-contrib-zwave-js).

@crxporter,
What's your thought about adding an 'Expert/Developer' checkbox in the driver config (which BTW @fabnavigator will be a config node when V9 lands).

Enabling this option, will expose the Driver instance in `global`
then 1 could do this in a function node.

```javascript
const ValueID = {}
const Node = 7
const Driver = global.get('ZWJS-DRIVER-MyConfig')
const Value = Driver.controller.nodes.get(Node).getvalue(ValueID)

return {
    payload: Value
}
```

In affect `ZWJS-DRIVER-MyConfig` exposes an untethered driver instance `MyConfig` is essentially the identifier for the config node (where the instance actually lives) - so it still supports multi networks

As this is direct access to the driver - its use is not for the unexperienced, therefore I would like to brand this as an Expert/Developer' option - default will be disabled, and using its functions is basically covered on the zwave js docs

OR, we could slim this down, and only expose certain (non-harmful) functions that can be used directly in function nodes.

```javascript
const ValueID = {}
const Node = 7
const ZWaveUtilities = global.get('ZWJS-DRIVER-UTILS-MyConfig')
const Value = ZWaveUtilities.ValueAPI.getValue(Node,ValueID)

return {
    payload: Value
}
```

I will need to test this of course, but more importantly do you think it can be helpful?
it will be an interesting/advanced feature to add - but not sure how best to execute/shape it 

Adding global functions this way is highly documented on the forums - so is not at all 'wrong'
the NR settings file also supports this mechanism .

```
 /** The following property can be used to set predefined values in Global Context.
     * This allows extra node modules to be made available with in Function node.
     * For example, the following:
     *    functionGlobalContext: { os:require('os') }
     * will allow the `os` module to be accessed in a Function node using:
     *    global.get("os")
     */
    functionGlobalContext: {
        // os:require('os'),
    },
```

Discuss..... 😄

**Reply by crxporter** [reply]
I wouldn't call it expert ... I'd just call it what it is. A tick box with "Save values DB to global context?"

It's very interesting and is almost exactly what Mr Fabulous Navigator was looking for. I think it's probably the "right way" to do this with a condition that it stays up-to-date whenever any zwave thing changes it goes into the global DB variable.

The next concern for a network the size of mine is how often will it change and how often will it be written to disk? That would be potentially 10 times the data (or more) than I currently keep in context storage. Writing it to disk as often as I do my current stores would be a lot. 

So now I'm getting to the next level ......... what if it's a new context store? I'm looking at the API docs and it looks like "Store Module API" would be awesome here. We wouldn't use the regular context store but there would be a zwave context store. We would pull it like `global.get('zwaveThing','zwaveStore')` then we know it's from your special zwave store. This works from the UI also (like for a change node) with a drop-down for which store to use.

If you can't tell by now... yes I think I like this idea.

**Reply by marcus-j-davies** [reply]
got you message.
I will investigate what's possible,

What I am hoping for is the ability to not have to store values our self - it starts getting messy.
so I will see if I can 'hijack' the `get `command in our custom store - to effectively offload the fetch with the values already in the driver instance, a kind of cloak around the driver using the storage API if you will - watch this space 

EDIT:
it would seem custom store API is not suitable -it needs to be added to `settings.js` - we shouldn't expect that type of alteration to be required ☹️.

it will need to use the normal store, presently I'm leaning on the Utility side set of functions, as apposed to exposing the full driver

**Reply by marcus-j-davies** [reply]
![image](https://user-images.githubusercontent.com/55892693/200310937-7cee0378-fc80-47e1-8d7b-bede4bce0586.png)

**Reply by crxporter** [reply]
>  a kind of cloak around the driver using the storage API if you will 

So you're hinting that we might have access to the whole zwave API as a function in the global context? Which brings us full circle to a request I made long, long ago to have some more "raw" access to zwave js? (JS objects can be functions, right?)

If you can't make a new store then please let *me* make a new store. I'll want this in RAM. Wherever we turn it on, we should be able to pick the store if our nodered instance has multiple globals... I sent you a screenshot (text) earlier from the last section of [this](https://nodered.org/docs/api/context/#store-module-api) nodered doc page. No idea if it'll help but it sure would be neat to have a zwave store.

But please don't name it *that*... it's so long... `ZWJS-DRIVER-UTILS-MyConfig`

**Reply by marcus-j-davies** [reply]
`ZWJS-DRIVER-UTILS-MyConfig` is an example of a key that will expose utility methods to be used in user functions.
this key needs to allow for mulit network - so will need to think about its name

**Reply by marcus-j-davies** [reply]
@fabnavigator @crxporter 

This is the type of interaction to expect - if this happens the way I want it.
Note: executing `setValue `or `pollValue` will still trigger the normal event payloads, as these are attached to global event handles inside the driver 

![image](https://user-images.githubusercontent.com/55892693/200337730-d8bcd0f7-b050-49be-a2e5-7c01874ea778.png)

**Reply by fabnavigator** [reply]
```
const ValueID = {}
const Node = 7
const Driver = global.get('ZWJS-DRIVER-MyConfig')
const Value = Driver.controller.nodes.get(Node).getvalue(ValueID)
```
Is exactly what I was looking for.

I like: ```OR, we could slim this down, and only expose certain (non-harmful) functions that can be used directly in function nodes.```. Maybe then we don't need to add a checkbox to enable this new functionality.

I wouldn't write any of this to disk, but keep it in memory. All of the cached data will be reloaded if we restart nodered.

P.S. The 'fab" in "fabnavigator" is short for "fabrication" as in an integrated circuit wafer fabrication facility. "navigator" is a reference to the old Netscape Navigator web browser. Fab Navigator is the name of an application I wrote in the late 90s.

**Reply by crxporter** [reply]
> P.S. The 'fab" in "fabnavigator" is short for "fabrication" as in an integrated circuit wafer fabrication facility. "navigator" is a reference to the old Netscape Navigator web browser. Fab Navigator is the name of an application I wrote in the late 90s.

I figured as much but I thought it was funnier to say fabulous - hope I didn't offend 😆

**Reply by marcus-j-davies** [reply]
Here is a sneak peak.


https://user-images.githubusercontent.com/55892693/200385017-a2475b9d-7a13-46e8-ab88-4db987b10142.mp4

**Reply by marcus-j-davies** [reply]
Do Note: This is still **very** early days and is subject to massive change

**Reply by marcus-j-davies** [reply]
> I wouldn't write any of this to disk, but keep it in memory

You wouldn't need to anyway - it will be available in function/change/switch nodes as and when you need it

**Reply by crxporter** [reply]
> I wouldn't write any of this to disk, but keep it in memory

Also just saw this comment from @fabnavigator - this would be a setting based on where the user stores the global context I believe.

@marcus-j-davies you know there can be multiple global stores, right? Probably need to consider that...

**Reply by marcus-j-davies** [reply]
There is till investigation to be done, Im still deciding how best to reveal the driver in this way.
There are areas that I need to not display as they are private APIs (_something) as messing with them could cause crashes.

**Reply by fabnavigator** [reply]
@crxporter I never gt offended when someone uses "fabulous" to refer to me.

**Reply by fabnavigator** [reply]
@marcus-j-davies, This looks like a great start. When can I test it out? 😁

**Reply by marcus-j-davies** [reply]
it will be sometime!
its a rewrite which will likely take months of work. 

The `global` object will have the following tooling.

![Screenshot 2022-11-07 at 19 21 00](https://user-images.githubusercontent.com/55892693/200396233-0dc6f391-f610-4760-8619-da2fd4856df5.png)

**Reply by fabnavigator** [reply]
Hi @marcus-j-davies, One more thought on this topic. It would be useful to be able to call a function that returned the timestamp of the last update to the value.

**Reply by marcus-j-davies** [reply]
Hi @fabnavigator 
Sure, will list it for consideration. Work on V9 is in full swing!

**Reply by fabnavigator** [reply]
Thank you. Right now I use context to save the time and state. That way I can make decisions based on how long ago something happned. (i.e., When the house-to-garage door opens, turn on the garage lights. If the garage door opens within 30 seconds of that event, turn off the garage lights because you are just pulling out the car, not hanging out in the garage).

**Reply by marcus-j-davies** [reply]
You do know V8 had the ability to fetch  last timestamps right 😆 

https://github.com/zwave-js/node-red-contrib-zwave-js/wiki/Driver-API#getLastEvents

**Reply by fabnavigator** [reply]
I probalby saw that at some point. I don't remember. In any case, I need to be able to access this information from within the function node that is making these decisions. The last timestamp goes hand-in-hand with the getValue information. Could getValue return an object with the value and the timestamp. That would be cool!

**Reply by marcus-j-davies** [reply]
> Could getValue return an object with the value and the timestamp

I like the thought off `getValue` response actually representing when the value was actually  updated!

**Reply by marcus-j-davies** [reply]
@fabnavigator @crxporter 

New UI structure!

Certain areas of management, used to open up their own UI, well these UI's are now collectively accessed via the cog menus.

Controller UI's
![Screenshot 2023-01-14 at 13 08 47](https://user-images.githubusercontent.com/55892693/212473167-de730f9d-c444-47a0-954a-4ceb98f69550.png)

Node UI's (CC's are still listed, and managed as before)
![Screenshot 2023-01-14 at 13 09 25](https://user-images.githubusercontent.com/55892693/212473218-20e36a46-4c86-4f2d-bedb-3819fabaf3c4.png)

**Reply by fabnavigator** [reply]
@marcus-j-davies
Have you compared mouse clicks and movment to do common tasks, between the old and new structures? I expect the old is less work to use, but I imagine the new structure is easier to maintain/enhance. And can you pleae flesh out the example so we can see what the node list will look like, in relation to the buttons? It's cryptic now to have to click twice to get the buttons to appear. Are the buttons always going to be visible?

**Reply by marcus-j-davies** [reply]
The buttons that are overlayed will be gone, and will be replaced with this Unified approach.

Example: to name and set the location of a node
(imagine there are 2 text boxes for name and location on the empty panel) and a save button

https://user-images.githubusercontent.com/55892693/212476104-017b18f0-f033-4d8d-9b0e-22e81e1eafe3.mov



Example of Controller information (actual data)
![Screenshot 2023-01-14 at 14 04 35](https://user-images.githubusercontent.com/55892693/212476034-62206b09-f83c-4d99-bffb-a9858a09bbba.png)

**Reply by fabnavigator** [reply]
I understood the dialog part to make the changes. What I can't picture is what the node list will look like. Will I see a cog under or above each node name?

**Reply by marcus-j-davies** [reply]
Initially, it would be a requirement to select the node, then select the cog at the bottom, But I may have a cog that follows you as you hover over each node in the list - watch this space

**Reply by fabnavigator** [reply]
I see. I'm good with selecting a node from the list and then seeing the cog as part of the node information below the list of nodes. I would vote against any hovering (if I get a vote).

**Reply by marcus-j-davies** [reply]
Agreed - will keep it as a cog at the bottom section

**Reply by marcus-j-davies** [reply]
@fabnavigator @crxporter 

Some sneak peaks



https://user-images.githubusercontent.com/55892693/213299542-20041ec4-e56b-4a04-bf82-c24d44e3e63f.mov

**Reply by fabnavigator** [reply]
@marcus-j-davies 

Looking good! Does the "esc" key cancel out of the dialogs, or do you have to hit the "X"? Also, the real-time log is cool, but how about it at the node level also? The messages sent/received by the selected node.

**Reply by marcus-j-davies** [reply]
The logging is direct from the driver.

You can filter the logs, to a specific node using the controller node config values (it's there already) but, it's still managed by the driver and not the module.

The modal dialogs, I'm still working on

**Reply by fabnavigator** [reply]
Is the default for the log to report on all nodes? If so, wouldn't it just be a matter of getting the log information as you are, but only display what pertains to the selectd node? If someone did set the controller to only log for a specific node, and you are on another node, you wouldn't get anything. I'm not sure if it could let you know that was the case.

**Reply by marcus-j-davies** [reply]
I'm not "getting" the log info in any packaged way.

It's tailing the drivers log file and feeding it to the UI using web sockets.

The driver can be be configured to only report on a select number of nodes, what that does I don't actually know (never used it) but it's affect will be sees in the live log view, as it's just tailing whatever log entries are being written.

Configuring the log settings, is on the controller node, in the current version. And most of the settings will remain, only the file path can't be customised on the new version currently.

If you don't specify any nodes on the log settings example: 7,5,8 - then it won't be filtered.

**Reply by fabnavigator** [reply]
I think I understand.

**Comment by marcus-j-davies** [comment]
What is peoples thought on this?

The topic will be used to contain the API and method of that API

example : `VALUE_API.setValue`
will execute `setValue`, on the `Value API`

the properties, will no longer be unnamed in an array, and will now be attached to property names.
`options` has always been optional (and will remain so)
 
```js
{
    topic: 'VALUE_API.setValue',
    payload: {
        valueId: {},
        nodeId: 345,
        value: 3,
        options:{}
    }
}
```


Using the old method,  the current way is:
```js
{
    payload: {
        mode: 'ValueAPI'
        method: 'setValue',
        node: 345,
        params: [<ValueID>,3,<Options>]
    }
}
```


I believe this makes for a cleaner, more structured  understanding.
it also allows better manipulation  with core nodes

**Reply by fabnavigator** [reply]
@marcus-j-davies 

I'm sorry, but I can't get too excited about a change that will break a bunch of my code.  Also, sometimes I use msg.topic, and now I will have to keep in mind that it has z-wave significance.

Is this change really going to be better in the long run?

**Reply by marcus-j-davies** [reply]
@crxporter also shot down the use of topics.

I guess I won't be using topics, the other changes may remain.

The internal structures/ code base has changed, and not in a small way.

I'll try to soften the blow as best as I can - but certain payload changes will be required, not to mention nodes needing to be redeployed.

**Reply by crxporter** [reply]
What other breaking changes are coming? Might be a good time to list those so we can brace ourselves.

Are the controllers going into config nodes? I think I remember hearing about that...

**Reply by fabnavigator** [reply]
@marcus-j-davies At this point in the maturity of the product, I wish you would consider deprecating things, not just changing something that breaks our automations (some of us are married and want to stay that way). That way we can have at least until the next major release to transition our code base to the new way to do something. In terms of testing the new things, maybe you could write something that translates the old way to do something to the new way. That way we would be running your new code, even thought we didn't modify our code.

**Reply by marcus-j-davies** [reply]
So what I'm getting, is that the request format isn't a welcomed change?

I will therefore look to keep the same format or warn that it should be updated at least.
One thing I can't avoid, and that is having to replace nodes with newer versions, as these will move to config nodes.

It will (hopefully) just be a drop-in-place and configure thing, this should only apply to your controller and device nodes.

I will soon raise a Draft PR for v9.0.0 - just to aid in documenting the changes. This will change overtime as I progress.

**Reply by marcus-j-davies** [reply]
These are the proposed changes to the new message format.
The old format will still be available during the entirety of v9.

CCAPI Example
```js
{
  payload: {
    cmd: 'CCAPI.InvokeCCAPI',
    nodeId: 345,
    commandClass : number | string,
    arguments: [0,0,0],
  }
}
```

ValueAP Example (Set Value)
```js
{
  payload: {
    cmd: 'ValuAPI.setValue',
    nodeId: 345,
    valueId: { }.
    SetValueOptions: { }, /* Optional */
    value: 34535
  }
}
```

And for good measure, obtaining the DB
```js
{
  payload: {
    cmd: 'ControllerAPI.getValueDB',
    nodes: [6,5,8] /* Optional */
  }
}
```

**Reply by fabnavigator** [reply]
Thank you. That will work for me. It would be great if the documentation included the old and new formats together, to make it easy for people to update their code.

In terms of newer versions of nodes. I just have a single instance of the controller node. What will happen when I upgrade? Will my old controller node just disappear?

**Reply by marcus-j-davies** [reply]
There will be a migration document like there  was to aid moving to v4 from v3 (below)
https://github.com/zwave-js/node-red-contrib-zwave-js/blob/main/APIChange.md

> What will happen when I upgrade? Will my old controller node just disappear?

The old controller will be marked as missing, and you just replace with the new node.
Therefore backup your keys!!!! the network its self will recover without out a hitch - as the database for your network wont change

**Reply by fabnavigator** [reply]
I just exported and downloaded the json for my controller node. Thanks for the heads-up.

**Comment by marcus-j-davies** [comment]
So...

A little update on new access points for the Various APIs.

all `payload`  objects will now have a `cmd` object, and this denotes what your doing, my aim here is to create a clear separation internally, and to aid in  payload  viewing

There will be 4 APIs - each serving a specific area with various methods for each
```
CONTROLLER
CC
VALUE
NODE

```

Examples

```js
/* Ping a node */
{
  payload: {
    cmd: 'NODE.ping',
    nodeId: 34
  }
}

/* Set a node name */
{
  payload: {
    cmd: 'NODE.setNodeName',
    nodeId: 34,
    value: 'My Node'
  }
}

/* Get Value DB for specific nodes */
{
  payload: {
    cmd: 'CONTROLLER.getValueDB',
    nodes: [34,25] /*Optional*/
  }
}

/* Update a value (Value API) */
{
  payload: {
    cmd: 'VALUE.setValue',
    nodeId: 52,
    valueId :{},
    value : 5352,
    setValueOptions:{} /*Optional*/
  }
}

/* Update a value (CC API) */
{
  payload: {
    cmd: 'CC.invokeCCAPI',
    commandClass: 'Configuration',
    method: 'set',
    nodeId: 52,
    endpoint: 2, /* Optional */
    args:[{"parameter": 4, "value": 0 "valueSize": 1}] /* Optional */
  }
}
```

Im still chewing around some bits - but you should hopefully get the idea

**Reply by fabnavigator** [reply]
It looks like "mode/method" is somewhat replaced with "cmd/method".

**Reply by marcus-j-davies** [reply]
yup.

`cmd` is now a pointer to the function that will get executed - a path if you will to what is being executed.

internally, it will take the left side (the API) and will branch off to the forest of methods in that API, and will execute the right side

**Reply by crxporter** [reply]
These are shortcuts to zwave js internals?

**Reply by marcus-j-davies** [reply]
Not 'shortcuts' per se - but named accureltey according to the internals of the underlining driver

**Reply by marcus-j-davies** [reply]
Another potential.

```js
const Message = {
    payload: {
        cmd: {
            api: 'VALUE',
            method: 'setValue'
        },
        cmdProperties: {
            nodeId: 34,
            valueId:{ },
            value: true,
            setValueOptions:{ }
        }
    }
}
```

**Reply by fabnavigator** [reply]
I like that better. More jsonly.

**Reply by marcus-j-davies** [reply]
Agreed!

**Comment by marcus-j-davies** [comment]
Thoughts! @fabnavigator @crxporter 


```js
/* You will (likely) be notified that the value has been updated on the device - this is hit and miss now, and is currently the norm */
const Message = {
    payload: {
        cmd: {
            api: 'VALUE',
            method: 'setValue'
        },
        cmdProperties: {
            nodeId: 34,
            valueId:{ },
            value: true,
            setValueOptions:{ }
        }
    }
}

/* You will get an output message with your deliveryReceipt, once the value has been committed to the device
    think of battery operated devices for example  - that wake up every so often
    deliveryReceipt - can be any type of value (number ,string, object)
*/
const Message = {
    payload: {
        cmd: {
            api: 'VALUE',
            method: 'setValue'
        },
        cmdProperties: {
            nodeId: 34,
            valueId:{ },
            value: true,
            setValueOptions:{ },
            deliveryReceipt: 5645465
        }
    }
}
```

**Reply by crxporter** [reply]
What's different between this way and just using force update?

```json
"forceUpdate":{"property":"currentValue"}
```

**Reply by crxporter** [reply]
oh. force update is for CC api and delivery receipt is for value api?

**Reply by marcus-j-davies** [reply]
`forceUpdate` is nothing to do with being notified that a value has _really_ been updated

**Reply by marcus-j-davies** [reply]
`forceUpdate` fetches the value (internally) after its been updated (Its CCAPI acting as if its the ValueAPI)

**Reply by crxporter** [reply]
Whatever we put as delivery receipt comes back with the confirmation message later? I see the benefit in that!

**Reply by marcus-j-davies** [reply]
Yes!!!

what you provide in `deliveryReceipt` - will be returned when the device ACK's the request

**Reply by marcus-j-davies** [reply]
Proof of delivery one would say

**Reply by crxporter** [reply]
Like a tracking number - an identifier for my request

**Reply by marcus-j-davies** [reply]
it will be any object you want to attach - and it will simply be returned say under an event of `DELIVERY_RECEIPT `

**Reply by fabnavigator** [reply]
@marcus-j-davies

I understand the concept of a delivery receipt, but I would like you to elaborate on your comment ```You will (likely) NOT be notified that the value has been updated on the device - this is hit and miss now, and is currently the norm```. As far as I know I'm getting notifications from the controller after I update values. If I wasn't, my stuff wouldn't be working.

**Reply by marcus-j-davies** [reply]
Yup - and most CC's (or devices) respond with their new value!  = `VALUE_UPDATED`
`VALUE_UPDATED` - has always been when the device responds with the new value.

BUT

Some CC's (or devices) can be lazy - in that, they will accept the new value - but not necessarily tell us (the driver will do what it can to get confirmation in these situations).

this new (optional) token allows you to know if the command has been processed by the device - whether or not it will respond with its new value - think of as 'proof of delivery'

Example : The Indicator CC DOES NOT produce a value `VALUE_UPDATED` - so using this token, you know that the indicator was set

**Reply by marcus-j-davies** [reply]
@fabnavigator  - I have sent you a video that I shared on Discord - demoing the token function

EDIT.
for some reason I can re-upload.

**Reply by fabnavigator** [reply]
So, if I use the token all the time, I can ignore the VALUE_UPDATED events and only process the DELIVERY_RECEIPT events?

**Reply by marcus-j-davies** [reply]
Technically yes! - the event is `TRACKING_TOKEN_RETURN`

and looks like this.

```js
{
  "event": "TRACKING_TOKEN_RETURN",
  "timestamp": 1674926618383,
  "eventBody": {
    "token": 100,
    "response": { }
  }
}
```

**Reply by fabnavigator** [reply]
That's good. I don't want to over complicate my flow with two events causing me to take the same action.

**Reply by marcus-j-davies** [reply]
you only get `TRACKING_TOKEN_RETURN` if you provide the token

**Reply by marcus-j-davies** [reply]
response is whatever the response is `true`, `false`, `error`, some other object

**Reply by marcus-j-davies** [reply]
I just realised I said:

_"You will (likely) NOT be notified that the value has been updated on the device - this is hit and miss now, and is currently the norm"_

I was meant to say

_"You will (likely) be notified that the value has been updated on the device - this is hit and miss now, and is currently the norm"_

**Reply by fabnavigator** [reply]
Thanks for that clarification. It makes more sense. Only getting TRACKING_TOKEN_RETURN sounds right.

**Reply by marcus-j-davies** [reply]
If using Tokens to track success, you will still get (most of the time) - the `VALUE_UPDATED` as this event is a by-product of the value updating if that make sense.

But my aim is to use the token approach as  a more solid confirmation for the end user, as these will happen once the command has completed - and whether or not `VALUE_UPDATED` will occur from it

**Reply by fabnavigator** [reply]
I understand. I misinterpreted your statement previously to think that you didn't value updated when you got tracking token return.

Now that I understand, I wonder whether tracking token return is very self-explanatory. I'm going to think about this little more.

**Reply by marcus-j-davies** [reply]
Im still playing around with the event name
`TRACKING_TOKEN_UPDATE` is another - but it will only ever have 1 update (when the operation is completed)

**Reply by fabnavigator** [reply]
This is something optional when you set a value, so could it be implemented as a couple of setValueOption items?
sendAck: true
-- Triggers the MSG_ACK event, which contains enough information for you to know which message has been acknowledged.
sendAckTag: object
-- If you need to get an object returned in the MSG_ACK event.

**Reply by marcus-j-davies** [reply]
> so could it be implemented as a couple of setValueOption items

I'm afraid not - `setValueOptions` is actually an object that is passed to the driver, and is not used by the node exclusively.
I should note that this token object now lives under `cmd` - `cmdProperties` should be exclusively left for driver values

```javascript
const Message = {
    payload: {
        cmd: {
            api: 'VALUE',
            method: 'setValue',
            trackingToken:100
        },
        cmdProperties: {
            nodeId: 34,
            valueId:{ },
            value: true,
            setValueOptions:{ },
        }
    }
}

```

I think Tracking Token is the right name. (but open to suggestions)
it tracks the executions of the command, Here is what it is doing

```javascript
Driver.nodes.get(nodeId).setValue(valueId, value setValueOptions)
    .then((Response) => {
        // Send Token back if it is provided 
    })
    .catch((Error) => {
        // Send Token back if it is provided 
    })

```

`setValue` returns on success (if battery operated, will wait until its awake) or error

I can change the object name and the event that it creates but the term ack.... might be pushing it a little?

**Reply by fabnavigator** [reply]
I didn't like ack much either, but that's what the z-wave sniffer uses.
Maybe you could explain the difference between what is "usually" returned for a setValue, to what is returned when the receipt is requested.

**Reply by marcus-j-davies** [reply]
`VALUE_UPDATED` 

 - Sensor is triggered - advertises a new value 
 - A device responds with a value update (as a result of you changing it)
    But this depends if the device is not lazy or you are using a CC that does not (by design) generate a value update
    (Indicator CC is one)
  
`TRACKING_TOKEN_RETURN`
 - A `setValue` (or any command) has finished executing
     - Either because it was able to talk to a device (waiting for it to wake if needs be) 
     - There was an error
     
     You will get the response as well as token value
     
`TRACKING_TOKEN_RETURN` = "The operation is complete, here is the response"

**Reply by marcus-j-davies** [reply]
`setValue` has NEVER returned anything in the module, the `VALUE_UPDATE` event is because the device just happens to respond to an update - but its not a direct return from your action - `VALUE_UPDATE` is a by-product

**Reply by marcus-j-davies** [reply]
@crxporter gave it this description.

_Like a tracking number - an identifier for my request_

And that is quite accurate

**Reply by fabnavigator** [reply]
I'm ok with those names.

**Reply by fabnavigator** [reply]
Maybe I was a bit thrown off because trackingToken is the only cmd property that is optional, but there is nothing indicating that.

**Reply by marcus-j-davies** [reply]
Yup, the payload is now to this format - which I think is pretty tidy

```js
const Message = {
    payload: {
        cmd: {
            api: 'Your Preferred API',
            method: 'Method on that API you want to call',
            trackingToken: any /* Optional*/
        },
        cmdProperties: {
          "Properties":"required for the Method you are calling"
        }
    }
}
```

**Reply by fabnavigator** [reply]
Would you want to indicate in the comment that it triggers the "TRACKING_TOKEN_RETURN" event?

**Reply by marcus-j-davies** [reply]
The Wiki material will document that, yeah 👍

**Comment by marcus-j-davies** [comment]
Another question.....

If I have a sensor with 2 channels (Lux, Temp)
would the below Association UI make sense?

![Screenshot 2023-01-29 at 13 08 27](https://user-images.githubusercontent.com/55892693/215328187-ea8b29ed-7143-4843-8d04-cf29596f37a4.png)

**Reply by fabnavigator** [reply]
Maybe not so much to me.  I would prefer source and target to local and remote. And I'm not sure what root device means.

**Reply by marcus-j-davies** [reply]
Root Device  = Endpoint 0

**Comment by fabnavigator** [comment]
Hi @marcus-j-davies,

I don't do many associations, but I got a new remote and want to set some up. It sure would be nice if I could associate to a node name instead of just the node number. I realize that I have write down all of the node numbers for the nodes I want to associate, and then go in and do the associations.

**Reply by marcus-j-davies** [reply]
HI @fabnavigator 

I have been preoccupied lately, so not had chance to progress.

I can't ask for a name - as that is not what the  protocol wants + plus not everyone names their nodes  - but I may instead prompt with a list of nodes when adding an association. I cant promise that the names will be listed in the associations UI though.

**Reply by fabnavigator** [reply]
@marcus-j-davies

That's no problem. We all get preoccupied.

A list w/o names probably isn;t worth a change, but a list of "node - name" or "node - " would be nice.

**Reply by marcus-j-davies** [reply]
It will be a list of Nodes with names (if you have set the name)

Once my time frees up - i'll continue with V9

**Comment by marcus-j-davies** [comment]
I cant remember who asked for this. But the genuine timestamp of a value is now available driver side.
so v9 will cash in on that. 

it might have been @rgerrans who requested it.
I currently "cheat" doing this - by storing timestamp table during run time, but this makes it more useful. 


https://github.com/zwave-js/node-zwave-js/pull/5556

**Reply by crxporter** [reply]
Ooh. I like this. More "real" timestamps!

**Reply by rgerrans** [reply]
Yep, that was me when I was working on how to do offline checks. I used the LAST_EVENTS_RESULT you setup to do my checks if a battery device is offline

Looking forward to the next version!!!

**Comment by marcus-j-davies** [comment]
Get Value DB

![Screenshot 2023-03-19 at 18 33 23](https://user-images.githubusercontent.com/55892693/226199254-325cd3a9-0434-416a-a896-c003278d786f.png)

Get Value

![Screenshot 2023-03-19 at 18 35 45](https://user-images.githubusercontent.com/55892693/226199271-166021bf-0554-4cf2-9269-bc0d665e048f.png)

**Comment by marcus-j-davies** [comment]
Hey everyone,

Just a quick update on V9 progress: it's still in my FOV, and I haven't forgotten about it. Apologies for the delay – But 2023 has been unexpectedly busy in both my professional and private life.

Despite the hectic schedule, I've been keeping V8 maintained, and V9 will soon be getting the attention it deserves. The upcoming release promises to be groundbreaking, thanks to the enhancements from the Z-Wave JS Driver, which I'm eager to include.

Thank you all for your patience and support!

**Comment by fabnavigator** [comment]
Hi @marcus-j-davies ,

I hope you have been well. It's been a while. Everything is still working great here with my 70 nodes. 

I noticed that v10.0.0 has been released. I'm currently on v9.1.3. Should I upgrade, or might some things break?

**Comment by marcus-j-davies** [comment]
Hi @fabnavigator 

Hope you are well?
I changed Job roles professionally, it's going great! but incredibly time consuming, time now is extremely tight.

All the big changes discussed here, didn't make it into V10 - due to my new role, and general private affairs - so haven't had time to implement them 😢

V10 is a breaking change, but only as far as the NodeJS Version.

See here : https://github.com/zwave-js/node-red-contrib-zwave-js/releases/tag/v10.0.0

What is discussed here - will be a complete set of new nodes, New UI, new message format, new way of using the set of modules, but that is now happening in V11 (Currently), its actually almost done.

But thankfully  - the network cache/state will remain unaffected (no need to re-do the network) - just have to re-add the Node RED nodes and re-engineer it.

Things have changed quite a bit in the driver its self, so can't avoid the number of breaking changes coming in V11

V11 Screenshots

<img width="1703" height="852" alt="image" src="https://github.com/user-attachments/assets/e18cdf64-90b2-4948-8f74-07a956396233" />

<img width="1681" height="874" alt="image" src="https://github.com/user-attachments/assets/60d6576e-f729-48c7-bee4-8adba881a053" />
